operator/builtin/input/file/reader.go

```diff
@@ -96,6 +97,49 @@ func (f *Reader) InitializeOffset(startAtBeginning bool) error {
 	return nil
 }
 
+// readHeaders will read a files headers when labelRegex is set
+func (f *Reader) readHeaders(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		file, err := os.Open(f.fileLabels.ResolvedPath)
+		if err != nil {
+			return fmt.Errorf("failed to open file: %s", err)
+		}
+		defer func() {
+			if err := file.Close(); err != nil {
+				f.Errorf("failed to close file: %s", err)
+			}
+		}()
+
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			byteMatches := f.fileInput.labelRegex.FindSubmatch(scanner.Bytes())
+			if len(byteMatches) != 3 {
+				// return early, assume this failure means the file does not
+				// contain anymore headers
+				return nil
+			}
+			matches := make([]string, len(byteMatches))
+			for i, byteSlice := range byteMatches {
+				matches[i] = string(byteSlice)
+			}
+			if f.Fingerprint.Labels == nil {
+				f.Fingerprint.Labels = make(map[string]string)
+			}
+			f.Fingerprint.Labels[matches[1]] = matches[2]
+		}
+
+		if err := scanner.Err(); err != nil {
+			return fmt.Errorf("scanner error: %s", err)
+		}
+	}
```

Comment:
I don't love that we're reimplementing file consumption from scratch here. I think we should try pretty hard to unify the code that handles this. 

I believe we can adapt `ReadToEnd` just slightly to do exactly what we want. This would also have the benefit of managing the offset appropriately for us. Roughly:

```go
type consumerFunc func(context.Context, []byte) error

func (f *Reader) ReadToEnd(ctx context.Context) {
	f.readFile(ctx, f.emit))
}

func (f *Reader) ReadHeaders(ctx context.Context) {
	f.readFile(ctx, f.readHeader)
}

func (f *Reader) readFile(ctx context.Context, consumer consumerFunc) {
	if _, err := f.file.Seek(f.Offset, 0); err != nil {
		f.Errorw("Failed to seek", zap.Error(err))
		return
	}
	fr := NewFingerprintUpdatingReader(f.file, f.Offset, f.Fingerprint, f.fileInput.fingerprintSize)
	scanner := NewPositionalScanner(fr, f.fileInput.MaxLogSize, f.Offset, f.fileInput.SplitFunc)

	// Iterate over the tokenized file, emitting entries as we go
	for {
		select {
		case <-ctx.Done():
			return
		default:
		}
		ok := scanner.Scan()
		if !ok {
			if err := getScannerError(scanner); err != nil {
				f.Errorw("Failed during scan", zap.Error(err))
			}
			break
		}
		if err := consumer(ctx, scanner.Bytes()); err != nil {
			f.Error("Failed to consume entry", zap.Error(err))
		}
		f.Offset = scanner.Pos()
	}
}

func (f *Reader) readHeader(ctx context.Context, msgBuf []byte) error {
	byteMatches := f.fileInput.labelRegex.FindSubmatch(msgBuf)
	if len(byteMatches) != 3 {
		// return early, assume this failure means the file does not
		// contain anymore headers
		return fmt.Errorf("...")
	}
	matches := make([]string, len(byteMatches))
	for i, byteSlice := range byteMatches {
		matches[i] = string(byteSlice)
	}
	if f.Fingerprint.Labels == nil {
		f.Fingerprint.Labels = make(map[string]string)
	}
	f.Fingerprint.Labels[matches[1]] = matches[2]
	return nil
}
```

Reply:
I just implemented this in the latest commits.